src/bin/dnstap-inject/main.rs

```diff
@@ -0,0 +1,205 @@
+        .module(module_path!())
+        .init()
+        .unwrap();
+
+    let file = File::open(opts.file).await?;
+    let socket = setup_socket(&opts.dns).await?;
+
+    let mut framed = Framed::new(file, FrameStreamsCodec {});
+    while let Some(frame) = framed.next().await {
+        match frame {
+            Ok(frame) => match frame {
+                Frame::ControlReady(_) => {
+                    bail!("Protocol error: READY frame not allowed here");
+                }
+                Frame::ControlAccept(_) => {
+                    bail!("Protocol error: ACCEPT frame not allowed here");
+                }
+                Frame::ControlStart(_) => {
+                    // XXX: Validate the content type embedded in the Start control frame payload.
+                }
+                Frame::ControlStop => {
+                    return Ok(());
+                }
+                Frame::ControlFinish => {
+                    bail!("Protocol error: FINISH frame not allowed here");
+                }
+                Frame::ControlUnknown(_) => {
+                    bail!("Protocol error: Unknown control frame");
+                }
+                Frame::Data(mut payload) => match dnstap::Dnstap::decode(&mut payload) {
+                    Ok(d) => {
+                        //let mut s = String::with_capacity(2048);
+                        process_dnstap_frame(&socket, opts.proxy, d).await?;
+                        //s.push_str("---\n");
+                        //std::io::stdout().write_all(s.as_bytes()).unwrap();
+                    }
+                    Err(e) => {
+                        bail!(
+                            "Protocol error: Decoding dnstap protobuf message: {}, payload: {}",
+                            e,
+                            hex::encode(&payload)
+                        );
+                    }
+                },
+            },
+            Err(e) => {
+                bail!("Protocol error: {}", e);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn setup_socket(server_addr: &SocketAddr) -> Result<UdpSocket> {
+    let local_addr: SocketAddr = if server_addr.is_ipv4() {
+        "0.0.0.0:0"
+    } else {
+        "[::]:0"
+    }
+    .parse()?;
+
+    let socket = UdpSocket::bind(local_addr).await?;
+    socket.connect(server_addr).await?;
+    debug!("Connected socket to DNS server: {:?}", &socket);
+    Ok(socket)
+}
+
+async fn process_dnstap_frame(socket: &UdpSocket, proxy: bool, d: dnstap::Dnstap) -> Result<()> {
+    if let Ok(dtype) = dnstap::dnstap::Type::try_from(d.r#type) {
+        if dtype == dnstap::dnstap::Type::Message {
+            if let Some(msg) = &d.message {
+                process_dnstap_message(socket, proxy, msg).await?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn process_dnstap_message(
+    socket: &UdpSocket,
+    proxy: bool,
+    msg: &dnstap::Message,
+) -> Result<()> {
+    if let Some(query_address) = &msg.query_address {
```

Comment:
is there any possibility that `msg.query_address` is not set when using this tool? I'm not super familiar with the dnstap format but it seems to be optional in the protobuf schema. If the dnstap message did not contain a a query address the message would be silently skipped, even though the query address is only used if the `--proxy` flag is supplied?

Should there be a warning message logged if this tool encounters a message with no query address, or it contains invalid data? What about if `msg.query_message` is not set?

Reply:
Yeah, it's permitted by the schema, and practically all fields in dnstap.proto are optional. But in practice servers should always be setting these fields.